docs/madr/decisions/017-mesh-proxy-template.md

```diff
@@ -0,0 +1,173 @@
+# MeshProxyTemplate
+
+* Status: accepted
+
+Technical Story: https://github.com/kumahq/kuma/issues/4737
+
+## Context and Problem Statement
+
+We need to revamp ProxyTemplate for the new policy matching.
+Proxy Template lets you adjust Envoy Config when Kuma policies do not provide a native way to configure thing you want.
+In a secured setup, it's only safe for mesh operators to operate Proxy Template. Otherwise, service owners could bypass any policy.
+
+There are a couple of problems with existing ProxyTemplate policy 
```

Comment:
I think we constantly have a problem when you want to patch something inside the list. For example, you want to alter something on Route, but Route is an item in VirtualHost.Routes array. So the only option we have today is to recreate the whole list of routes

Reply:
also, do we want to take into account https://github.com/kumahq/kuma/issues/5209?

Reply:
We can, but I don't think it's required for the new policy rollout.

For the lists, you are right. For all patches we use proto Merge functionality. It replaces the lists by default. It also cannot handle embedded TypedConfig like transport socket match in Cluster.

One of the idea that comes to my mind is to introduce additional JSON patch modification.
Unfortunately, it requires serializing Envoy config to JSON and deserializing it again, so there is definitely quite a big overhead. However, it could save us at the perf cost. We could state in docs that use it as a last resort

Reply:
Agreed to add Routes and JSONPatch